Sources/Swift/Tools/SentryViewPhotographer.swift

```diff
@@ -5,28 +5,65 @@ import CoreGraphics
 import Foundation
 import UIKit
 
+protocol ViewRenderer {
+    func render(view: UIView) -> UIImage
+}
+
+class DefaultViewRenderer: ViewRenderer {
+    func render(view: UIView) -> UIImage {
+        let image = UIGraphicsImageRenderer(size: view.bounds.size).image { _ in
+            view.drawHierarchy(in: view.bounds, afterScreenUpdates: false)
+        }
+        return image
+    }
+}
+
 @objcMembers
 class SentryViewPhotographer: NSObject, SentryViewScreenshotProvider {
-    
     static let shared = SentryViewPhotographer()
-    
     private let redactBuilder = UIRedactBuilder()
+
+    var renderer: ViewRenderer
+        
+    init(renderer: ViewRenderer) {
+        self.renderer = renderer
+        super.init()
+    }
+    
+    private convenience override init() {
+        self.init(renderer: DefaultViewRenderer())
+    }
         
     func image(view: UIView, options: SentryRedactOptions, onComplete: @escaping ScreenshotCallback ) {
-        let image = UIGraphicsImageRenderer(size: view.bounds.size).image { _ in
-            view.drawHierarchy(in: view.bounds, afterScreenUpdates: false)
-        }
+        let image = renderer.render(view: view)
         
-        let redact = redactBuilder.redactRegionsFor(view: view, options: options)
+        let redact = redactBuilder.redactRegionsFor(view: view, options: options).reversed()
         let imageSize = view.bounds.size
         DispatchQueue.global().async {
             let screenshot = UIGraphicsImageRenderer(size: imageSize, format: .init(for: .init(displayScale: 1))).image { context in
+                
+                context.cgContext.addRect(CGRect(origin: CGPoint.zero, size: imageSize))
+                context.cgContext.clip(using: .evenOdd)
+                
                 context.cgContext.interpolationQuality = .none
                 image.draw(at: .zero)
                 
                 for region in redact {
-                    (region.color ?? UIImageHelper.averageColor(of: context.currentImage, at: region.rect)).setFill()
-                    context.fill(region.rect)
+                    context.cgContext.saveGState()
+                    context.cgContext.concatenate(region.transform)
+                    
+                    let rect = CGRect(origin: CGPoint.zero, size: region.size)
+                    switch region.type {
+                    case .redact:
+                        (region.color ?? UIImageHelper.averageColor(of: context.currentImage, at: rect)).setFill()
+                        context.fill(rect)
+                        context.cgContext.restoreGState()
+                    case .clip:
+                        context.cgContext.addRect(context.cgContext.boundingBoxOfClipPath)
+                        context.cgContext.addRect(rect)
+                        context.cgContext.restoreGState()
+                        context.cgContext.clip(using: .evenOdd)
```

Comment:
The goal of clip regions is to clip out any other drawing in the context, which is the reverse functionality of the clip function. 
Lines 45/46 are creating a clip region that cover the entire image, and because of the evenOdd + boundingBoxOfClipPath this new clip regions are used to exclude the drawing.

Reply:
I think a comment here explaining what we're doing would help a lot, also took me a while to understand 🙂 
maybe even an ascii drawing like this (chatgpt'd):

```
+------------------------+
|          \             |
|           \            |
|            \           |
|    +---------------+   |
|    |         \     |   |
|    |          \    |   |
|    |   rect    \   |   |   <-- we want this "rect" parts that are not overlapped by the line "\"  
|    +---------------+   |
|                  \     |
| boundingBoxOfClipPath  |
+------------------------+
```